### PR TITLE
[Cocoa] Improve support for VP codecs in TrackConfiguration

### PR DESCRIPTION
**Pull request by jernoble**
#### d9d895830a49d1c4fa6aaef5ab048c22bf47a176
<pre>
[Cocoa] Improve support for VP codecs in TrackConfiguration
<a href="https://bugs.webkit.org/show_bug.cgi?id=253789">https://bugs.webkit.org/show_bug.cgi?id=253789</a>
rdar://106609888

Reviewed by Eric Carlson.

VP8 and VP9 media content are commonly carried in .webm files, which have mostly optional support
for carrying codec configuration data in the media metadata. This means that at the time where track
data is parsed from the WebM file, video or audio configuration data may not yet exist.

Add support for noticing that the format description of a WebM track has &quot;changed&quot; when the first
keyframe is parsed, and add support for notifying the SourceBufferPrivate when this happens.

When these format descriptions change, push the new values up into the previously parsed
TrackPrivates, and use the values to recreate the track&apos;s configuration.

* LayoutTests/media/media-source/content/test-vp9-profile0-manifest.json: Added.
* LayoutTests/media/media-source/content/test-vp9-profile0.webm: Added.
* LayoutTests/media/media-source/content/test-vp9-profile2-manifest.json: Added.
* LayoutTests/media/media-source/content/test-vp9-profile2.webm: Added.
* LayoutTests/media/media-source/media-source-webm-configuration-change-expected.txt: Added.
* LayoutTests/media/media-source/media-source-webm-configuration-change.html: Added.
* Source/WebCore/Modules/webcodecs/VideoColorPrimaries.idl:
* Source/WebCore/Modules/webcodecs/VideoMatrixCoefficients.idl:
* Source/WebCore/Modules/webcodecs/VideoTransferCharacteristics.idl:
* Source/WebCore/platform/MediaSample.h:
* Source/WebCore/platform/graphics/AudioTrackPrivate.h:
(WebCore::AudioTrackPrivate::setFormatDescription):
* Source/WebCore/platform/graphics/PlatformVideoColorPrimaries.h:
* Source/WebCore/platform/graphics/PlatformVideoMatrixCoefficients.h:
* Source/WebCore/platform/graphics/PlatformVideoTransferCharacteristics.h:
* Source/WebCore/platform/graphics/VP9Utilities.cpp:
(WebCore::isValidVPProfile):
(WebCore::isValidBitDepth):
(WebCore::isValidRange):
(WebCore::isValidChromaSubsampling):
(WebCore::parseVPCodecParameters):
(WebCore::createVPCodecParametersString):
* Source/WebCore/platform/graphics/VP9Utilities.h:
* Source/WebCore/platform/graphics/VideoTrackPrivate.h:
(WebCore::VideoTrackPrivate::setFormatDescription):
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::didUpdateFormatDescriptionForTrackId):
(WebCore::SourceBufferPrivateAVFObjC::append):
* Source/WebCore/platform/graphics/cocoa/AudioTrackPrivateWebM.cpp:
(WebCore::AudioTrackPrivateWebM::codec const):
(WebCore::AudioTrackPrivateWebM::sampleRate const):
(WebCore::AudioTrackPrivateWebM::numberOfChannels const):
(WebCore::AudioTrackPrivateWebM::setFormatDescription):
(WebCore::AudioTrackPrivateWebM::updateConfiguration):
* Source/WebCore/platform/graphics/cocoa/AudioTrackPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/SourceBufferParser.h:
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
(WebCore::WebMParser::OnEbml):
(WebCore::WebMParser::VideoTrackData::consumeFrameData):
(WebCore::WebMParser::AudioTrackData::consumeFrameData):
(WebCore::WebMParser::formatDescriptionChangedForTrackData):
(WebCore::SourceBufferParserWebM::formatDescriptionChangedForTrackID):
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.h:
(WebCore::WebMParser::Callback::formatDescriptionChangedForTrackID):
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.mm:
(WebCore::createVideoInfoFromVPCodecConfigurationRecord):
* Source/WebCore/platform/graphics/cocoa/VideoTrackPrivateWebM.cpp:
(WebCore::VideoTrackPrivateWebM::setFormatDescription):
(WebCore::VideoTrackPrivateWebM::codec const):
(WebCore::VideoTrackPrivateWebM::width const):
(WebCore::VideoTrackPrivateWebM::height const):
(WebCore::VideoTrackPrivateWebM::colorSpace const):
(WebCore::VideoTrackPrivateWebM::updateConfiguration):
* Source/WebCore/platform/graphics/cocoa/VideoTrackPrivateWebM.h:

Originally-landed-as: 7455323f34a9. rdar://106609888
Canonical link: <a href="https://commits.webkit.org/263919@main">https://commits.webkit.org/263919@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/4185ac2d7beff2ddef285756e95b74a5e680cdfe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6123 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6308 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6491 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7686 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6470 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6123 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6527 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6261 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/9351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6233 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6251 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5542 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7748 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3727 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5522 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/13430 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5592 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/5600 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/7837 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6068 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/4968 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5489 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/5451 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1447 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9631 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5857 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm [hunk 1]
```diff
@@ -518,6 +518,19 @@ static void bufferWasConsumedCallback(CMNotificationCenterRef, const void* liste
     didReceiveSample(WTFMove(mediaSample));
 }
 
+void SourceBufferPrivateAVFObjC::didUpdateFormatDescriptionForTrackId(Ref<TrackInfo>&& formatDescription, uint64_t trackId)
+{
+    if (is<VideoInfo>(formatDescription)) {
+        auto result = m_videoTracks.find(AtomString::number(trackId));
+        if (result != m_videoTracks.end())
+            result->value->setFormatDescription(downcast<VideoInfo>(formatDescription.get()));
+    } else if (is<AudioInfo>(formatDescription)) {
```
Comment (eric-carlson) [inline Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm]: can you use an early return here?

```
    return;
}

if (is<AudioInfo>(format…
```

File: Source/WebCore/platform/graphics/cocoa/AudioTrackPrivateWebM.cpp [hunk 2]
```diff
@@ -106,6 +108,76 @@ std::optional<MediaTime> AudioTrackPrivateWebM::discardPadding() const
     return m_discardPadding;
 }
 
+String AudioTrackPrivateWebM::codec() const
+{
+    if (m_formatDescription) {
+        if (!m_formatDescription->codecString.isEmpty())
+            return m_formatDescription->codecString;
+        return String::fromLatin1(m_formatDescription->codecName.string().data());
```
Comment (eric-carlson) [inline Source/WebCore/platform/graphics/cocoa/AudioTrackPrivateWebM.cpp:116]: Is it worth setting   `m_formatDescription`?

File: Source/WebCore/platform/graphics/cocoa/VideoTrackPrivateWebM.cpp [hunk 3]
```diff
@@ -90,12 +100,17 @@ int VideoTrackPrivateWebM::trackIndex() const
 
 String VideoTrackPrivateWebM::codec() const
 {
+    if (m_formatDescription) {
+        if (!m_formatDescription->codecString.isEmpty())
+            return m_formatDescription->codecString;
+        return String::fromLatin1(m_formatDescription->codecName.string().data());
```
Comment (eric-carlson) [inline Source/WebCore/platform/graphics/cocoa/VideoTrackPrivateWebM.cpp:106]: Ditto

**Review by eric-carlson — APPROVED** [review]

File: Source/WebCore/platform/graphics/cocoa/AudioTrackPrivateWebM.cpp (hunk 2 above)
Comment (jernoble) [reply Source/WebCore/platform/graphics/cocoa/AudioTrackPrivateWebM.cpp:116]: I'm not sure I understand the question?

**Pull request merged**